### DISCUSSION TITLE
Azure B2C Authentication

### DISCUSSION THREAD

**Discussion by ADefWebserver** [oqtane/oqtane.framework#2186] [Q&A] [answered]
I can't seem to get Azure B2C Authentication to work.

This is my set-up in Azure B2C:
![image](https://user-images.githubusercontent.com/1857799/167303670-7bc5290a-b285-4a78-b67c-590664d498cc.png)

This is my set-up in Oqtane (notice I have a specific Custom user "b2c_1a_signup_signin_aad" flow specified):

![image](https://user-images.githubusercontent.com/1857799/167303730-1a237324-aeab-48d5-9646-7599cff93dd3.png)

When it gets to this line it throws an error:

![image](https://user-images.githubusercontent.com/1857799/167303783-d753b6b1-cbd5-4d5e-ae81-6845d070333a.png)

![image](https://user-images.githubusercontent.com/1857799/167303830-467d6b9f-13ae-4377-9a00-853f59797ae7.png)

The error is thrown on this call: `new ClaimsPrincipal(identity)` but I can't seem to debug that call.

**Comment by sbwalker** [comment]
@adefwebserver the UI is supposed to include default values for scopes for OIDC - did you remove them? Scopes are important in OAuth as they indicate the permissions you are requesting. Also did you try with PKCE enabled?

**Comment by ADefWebserver** [comment]
@sbwalker - I tried it with the scopes you have here (and with **PKCE** enabled and disabled):
https://youtu.be/Yvp_PLvJqqk?t=988

![image](https://user-images.githubusercontent.com/1857799/167426967-7c927d59-1125-472b-81f4-6c18736a597c.png)

I still get the same error :(
I also tried the authority you used but that didn't work for me either (and I need to specify the Azure B2C Flow).

**Reply by sbwalker** [reply]
I will investigate further

**Reply by sbwalker** [reply]
So I just tested my own Azure B2C tenant just to ensure that no recent code changes have affected the External Login behavior - and it worked fine for me. So this would lead me to believe that there is a difference in your Azure B2C configuration when compared to mine. I have an Azure B2C tenant (oqtane.onmicrosoft.com) where I have a single User Flow defined with the following properties:

![image](https://user-images.githubusercontent.com/4840590/167458146-3f66a983-3d00-4981-a8e6-13a6bab162dd.png)

For my authority I specify: https://login.microsoftonline.com/oqtane.onmicrosoft.com/v2.0 and I entered my Client ID/Secret and used the default Scopes (with PKCE disabled). Note that I also had to add a user to my tenant to allow me to login to it with that email address. 

The event log message you posted above is being logged by the Remote Failure event. I am curious what error is being thrown on the line you highlighted in the screenshot above where the principal is being assigned - as it is likely a different error. I am also curious if you are able to step through the ValidateUser method and see what is happening.

**Reply by ADefWebserver** [reply]
@sbwalker - I think the issue is that I have multiple User Flows and I _must_ be able to specify a Flow.  If you can actually specify a Flow in your example I think that will allow you to diagnose the issue? 

Also, I stepped through the code several times and did not encounter _any_ problems or errors in validate user. 

Literally the only error is thrown on this call: `new ClaimsPrincipal(identity)`.

**Comment by sbwalker** [comment]
@ADefWebserver and what is the actual error thrown on ```new ClaimsPrincipal(identity)``` - the error message you posted above from the event log is from the OnRemoteFailure event - not the OnTokenValidated event.

**Reply by ADefWebserver** [reply]
@sbwalker - When debugging, I get to that line (you can't debug inside it) and then it errors out. I was hoping to debug to find out exactly-what-the-problem-is but I hit that line and 'boom' :(

**Reply by sbwalker** [reply]
You could put a Try...Catch around that line to catch the error. I am curious what your App Registration settings look like in Azure B2C

![image](https://user-images.githubusercontent.com/4840590/167498018-28b8b36a-0d0c-4ae6-bf0c-2907a72008ba.png)

**Reply by sbwalker** [reply]
@ADefWebserver it is really hard to find documentation for how to use Azure B2C with standard OIDC. This article indicates that there is support for policy parameter that can be sent to Azure B2C:

https://github.com/uglide/azure-content/blob/master/articles/active-directory-b2c/active-directory-b2c-reference-policies.md

I think your assumption is correct that my scenario works because I only have a single user flow - so Azure B2C most likely uses it by default. However if you have multiple user flows then you must be explicit. I will probably need to add another option to the OpenId Connect settings - however I do not want to include anything which is specific to an identity provider like Azure B2C so I will need to figure out how to make this a generic option.

**Reply by ADefWebserver** [reply]
> I will probably need to add another option to the OpenId Connect settings - however I do not want to include anything which is specific to an identity provider like Azure B2C so I will need to figure out how to make this a generic option.

Understood :)

My App Registration settings look like yours (see my screen shot in my first post on this thread)

**Reply by ADefWebserver** [reply]
@sbwalker - I finally had a chance to put a try/catch around the call and you're right it does get past that line. It fully exits the method and then steps into the error logging. Meaning it never enters the catch section of the try/catch I created.

![image](https://user-images.githubusercontent.com/1857799/167523270-6964f75d-8dfb-4693-9587-6600aef81aca.png)

**Reply by sbwalker** [reply]
@ADefWebserver so after more research it turns out that no code changes were required in order to get Azure B2C user flows working... it just required a very specific configuration:

Authority: https://{tenant}.b2clogin.com/tfp/{tenant}.onmicrosoft.com/{policy}/v2.0
(in my case it is https://oqtane.b2clogin.com/tfp/oqtane.onmicrosoft.com/b2c_1_login/v2.0)
Scopes: openid,profile,email,{ClientID} <- this is your GUID
Email Claim: emails

Please give this a try and let me know if it works for you.

**Reply by ADefWebserver** [reply]
@sbwalker - Got it. Give me two days or so. Thanks

**Reply by ADefWebserver** [reply]
@sbwalker - It worked :) These are the settings that worked for me:
![image](https://user-images.githubusercontent.com/1857799/167738458-fe4446d2-c9e9-4141-9545-e312f95440e7.png)

**Reply by sbwalker** [reply]
I am glad it worked for you. It's a bit frustrating that the documentation for configuring Azure B2C on OIDC is so poor. I also do not understand why I had to specify a different Email Claim type in my installation.

**Reply by ADefWebserver** [reply]
@sbwalker - Can you post another reply with the solution so I can mark it as the answer? Thanks

**Comment by sbwalker** [comment, marked as answer]
@ADefWebserver (reposting) so after more research it turns out that no code changes were required in order to get Azure B2C user flows working... it just required a very specific configuration:

Authority: https://{tenant}.b2clogin.com/tfp/{tenant}.onmicrosoft.com/{policy}/v2.0
(in my case it is https://oqtane.b2clogin.com/tfp/oqtane.onmicrosoft.com/b2c_1_login/v2.0)
Scopes: openid,profile,email,{ClientID} <- this is your GUID

**Reply by kevinchle** [reply]
I have a hard time trying to get OpenID to work.  I tried OAuth and getting close but still got the ERROR _ AADB2C99059: The supplied request must present a code_challenge.  There is so little documentation about how to configure External Login.
